Documentation on grid layout

pjlee11 opened the pull request:
Part of #841

_Documents the UX grid layout and product requirements for the articles project. Also details our CSS grid solution to meet the the layout and requirements_

The [preview view](https://github.com/BBC-News/simorgh/blob/142149221b6d60200d1de0660b340fd8471aff09/src/app/lib/README.md) should be easiest for reviewing this documentation 

- ~[ ] Tests added for new features~
- ~[ ] Test engineer approval~

Commit:
documentation on our layout designs and grid solution

Commit:
grammar change

Commit:
correct markdown

Commit:
Add screenshots/Figures

Commit:
markdown syntax

Commit:
more details on "additional content"

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,94 @@ Here are the font sizes in `px`s:
+- of 8 columns the article body content spans the central 6
+- *See Figure 4*
+
+**Between 600px - 1007px**
```
This is a repetition of line 63

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,94 @@ Here are the font sizes in `px`s:
+
+## Problems
+
+As we want content to go "full bleed" we cannot use max-width to define the 1008|1280px column layouts. Also as full bleed spans to the very edge of the viewport each indivdual component will need to know about the layout margins.
```
`indivdual` => `individual`

And - _would_ we want individual components to know about the layout margins? Feels like components should be dumb, ie "you have this much space to fill". I think that's probably what you mean already, just reads ambiguously to me!

pjlee11 replied:
I think the wording here is misleading so I'll clarify - The grid item (in this case a row) will need to know about the layout margins, the component itself shouldn't. I'll change this to be less ambiguous

Commit:
bad copy paste on titles

Commit:
individual*

Commit:
component wrapper knows about layout margins

Commit:
Rewording content & updating codepen/figure links

- Links to a different codepen of the CSS grid solution example and updates the matching Figure
- Rewords "components knowing about margins" to "rows knowing about margins"

Commit:
Pretty much re-write the grid documentation

- logically structures the documentation so it specifies the UX grid, then the article UX designs, the the technical implementation
- utilises tables to help explain the various breakpoint difference
- adds a TL;DR for a short overview

Commit:
Merge branch 'latest' into documentation-on-grid-layout

pjlee11 commented:
Adding alpha2.0 label as I consider this tech debt from alpha1.0

Commit:
Merge branch 'latest' into documentation-on-grid-layout

bcmn commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,91 @@ Here are the font sizes in `px`s:
+| :-------------- | :------ | :----------------------- | 
+| 0px - 399px     | 1       | 1 of 1                   | 
+| 400px - 599px   | 6       | 4 of 6                   | 
+| 400px - 599px   | 6       | 4 of 6                   | 
```
This row seems to be duplicated -- & unless I'm misunderstanding what's being communicated (I believe you're talking about the UX grid, not the CSS grid), I think the UX designs show the design at 0px - 399px & 400px - 599px to both be 6 of 6 columns?

This seems to be in line with enabling the grids on the [375px designs](https://zpl.io/an98YYl)

Commit:
Updates

- Remove duplicated table row in article layout 
- Add figure 6

Commit:
Removes the 1 column under 400px

This definition is no longer required

bcmn committed:
Merge branch 'latest' into documentation-on-grid-layout

bcmn reviewed (dismissed):
Looks great -- a really good description to have, thanks!

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+- UX margins are exclusive of the UX column definition
+- We have a requirement to allow some items to break out of the grid
+- To achieve these requirement we've used CSS grid to define an additional column either side of the UX grid
+- 
```
Delete?

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+* [Statically defined columns for above 1280px example](https://codepen.io/phil-lee/full/zmjaMO/)
+
+### Figure 1 - Article grid layout at 300px
+<img width="315" alt="Screenshot showing the article layout at 300px" src="https://user-images.githubusercontent.com/7791726/47848179-69e96780-ddc5-11e8-9f9d-b5e3f2fb7cca.png">
```
Any reason why this isn't in Markdown?

ChrisBAshton replied:
Oh, to limit the rendered size 👍

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+| 600px - 1007px  | 6       | 4 of 6                   | 
+| 1008px - 1280px | 8       | 6 of 8                   | 
+| 1280px +        | 10      | 6 of 10                  | 
+* content contained within the grid, therefore not "full bleed"
```
```suggestion
\* content contained within the grid, therefore not "full bleed"
```

Escape the `*` to have it render as an asterisk.

ChrisBAshton replied:
Hmm this doesn't appear to have gone through! (Might be a bug with the new GitHub 'suggest' feature)

ChrisBAshton replied:
Oops - was just looking at the wrong commit

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+Below 1008px we define all columns to be dynamic. 
+
+Above 1007px we statically define the widths of the columns due to the fact they are limited in the UX definition, which then allows us to have a dynamic column either side of the UX grid. 
+  EG: In CSS grid values we can specific `grid-template-columns: 1fr repeat(10, minmax(0, ${staticColWidth})) 1fr;` where the `1fr`s are dymanic and the columns and `repeat()` are the static width columns.
```
```suggestion
  EG: In CSS grid values we can specific `grid-template-columns: 1fr repeat(10, minmax(0, ${staticColWidth})) 1fr;` where the `1fr`s are dynamic and the columns and `repeat()` are the static width columns.
```

-> `dynamic`

dr3 commented on src/app/lib/README.md:
```diff
@@ -16,13 +16,13 @@ Here are the font sizes and line heights in `em`s, assuming a base font size of
 | Body Copy    | 0.9375        | 1.25            | 1             | 1.5             | 1             | 1.375           | Body copy (p)                                                    |
 | Pica         | 0.9375        | 1.25            | 1             | 1.375           | 1             | 1.25            | Headline for single onward journey                               |
-| Long Primer  | 0.9375        | 1.125           | 0.9375        | 1.125            | 0.875         | 1.125           | Image caption & Video caption & Onward journey timestamps        |
+| Long Primer  | 0.9375        | 1.125           | 0.9375        | 1.125           | 0.875         | 1.125           | Image caption & Video caption & Onward journey timestamps        |
```
Minor capitalization inconsistency with line 22

pjlee11 replied:
I haven't manually changed this, the editor is removing empty whitespace, but I'll edit any who

dr3 commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+| 600px - 1007px  | 6       | 4 of 6                   | 
+| 1008px - 1280px | 8       | 6 of 8                   | 
+| 1280px +        | 10      | 6 of 10                  | 
+* content contained within the grid, therefore not "full bleed"
```
This asterisk gets converted into a bullet point

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+| Breakpoints     | Columns           | Gutters   | Margins          | Figure |
+| :-------------- | :---------------- | :-------- | :--------------- | :----- |
+| 0px - 399px     | 6 + 2 of 0px      | 0 of 8px  | 2 of 8px         | 1      |
+| 400px - 599px   | 6 + 2 of 8px*     | 5 of 8px  | 2 of 16px        | 2      |
```
Don't think this needs an asterisk - you have explained three of the `Columns` rows below.

Do we need the Figure column? (I'd just remove)

pjlee11 replied:
I think I was being overly explicit but agree it's overkill, will update

dr3 commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+* [Investigation PR](https://github.com/BBC-News/simorgh/pull/824)
+* [Statically defined columns for above 1280px example](https://codepen.io/phil-lee/full/zmjaMO/)
+
+### Figure 1 - Article grid layout at 300px
```
This image is a bit misleading and doesnt show the 6 columns at the smaller widths, unlike Figure 2 which also has 6 columns but shows them, image implies <399 only has 1 column

pjlee11 replied:
Good spot, this is a hangover from a definition UX column breakpoint definition - will update

ChrisBAshton commented on src/app/lib/README.md:
```diff
@@ -41,3 +41,92 @@ Here are the font sizes in `px`s:
+
+`6 + 2 of 0px` - is based on having a 0px column either side of UX  grid so we can use CSS grid-gaps as the UX margin while allowing a grid item to be able to go "full bleed". 
+`6 + 2 of 8px*` - between 400px and 599px we have an 8px column either side of the UX grid so that the grid-gap (8px) and additional column (8px) combined meet the 16px UX margin.
+`0 of 8px` - under 400px we have only 1 column and therefore no gutters.
```
Looks like you intended this to be in a list (would read better that way)

```suggestion
* `6 + 2 of 0px` - is based on having a 0px column either side of UX  grid so we can use CSS grid-gaps as the UX margin while allowing a grid item to be able to go "full bleed". 
* `6 + 2 of 8px` - between 400px and 599px we have an 8px column either side of the UX grid so that the grid-gap (8px) and additional column (8px) combined meet the 16px UX margin.
* `0 of 8px` - under 400px we have only 1 column and therefore no gutters.
```

ChrisBAshton committed:
typo

Co-Authored-By: pjlee11 <phil.lee.ext@bbc.co.uk>

ChrisBAshton committed:
asterisk not bullet point

Co-Authored-By: pjlee11 <phil.lee.ext@bbc.co.uk>

ChrisBAshton committed:
Markdown as list

Co-Authored-By: pjlee11 <phil.lee.ext@bbc.co.uk>

Commit:
review updates and fix conflict from origin

Commit:
375px

dr3 commented on src/app/lib/README.md:
```diff
@@ -32,12 +32,108 @@ Here are the font sizes in `px`s:
+
+- `6 + 2 of 0px` - is based on having a 0px column either side of UX grid so we can use CSS grid-gaps as the UX margin while allowing a grid item to be able to go "full bleed".
+- `6 + 2 of 8px` - between 400px and 599px we have an 8px column either side of the UX grid so that the grid-gap (8px) and additional column (8px) combined meet the 16px UX margin.
+- `0 of 8px` - under 400px we have only 1 column and therefore no gutters.
```
Im sure you don't mean this exactly, but `under 400px we have only 1 column` doesn't sound correct as the table shows 6 + 2 columns. Might need to be more specific

bcmn replied:
I believe the UX requirement is one column, but for us the implementation is such an edge case that we don't do anything for it. It'd be good to document that fact, though

pjlee11 replied:
As Ben said, UX actually have a breakpoint where 1 column should be implemented but we currently don't have a use case/ support for browsers under 240px. I previously was told it was 400px hence this wording but I'll address this and document the breakpoints.

dr3 replied:
Yeah that makes sense, just concerned someone outside the team might get confused by css columns and UX columns, so we could be clearer

Commit:
update 0px - 399px breakpoint and add final note section

dr3 reviewed (approved):
👍 LGTM, great doc writing

ChrisBAshton reviewed (approved):
Looks good 👍

Commit:
Merge branch 'latest' into documentation-on-grid-layout
